File: python/protein-translation/protein_translation.py
CODONS = {
    "AUG": "Methionine", "UUU": "Phenylalanine", "UUC": "Phenylalanine",
    "UUA": "Leucine", "UUG": "Leucine", "UCU": "Serine", "UCC": "Serine",
    "UCA": "Serine", "UCG": "Serine", "UAU": "Tyrosine", "UAC": "Tyrosine",
    "UGU": "Cysteine", "UGC": "Cysteine", "UGG": "Tryptophan",
    "UAA": "STOP", "UAG": "STOP", "UGA": "STOP",
}


def proteins(strand):
    proteins = []

    for codon in chunks(strand, 3):
        if CODONS[codon] == "STOP":
            break
        proteins.append(CODONS[codon])

    return proteins


def chunks(s, n):
    for i in range(0, len(s), n):
        yield s[i:i + n]

File: haskell/protein-translation/src/ProteinTranslation.hs
module ProteinTranslation
  ( proteins
  ) where

import           Control.Applicative            ( liftA2 )
import           Data.List.Split                ( chunksOf )

proteins :: String -> Maybe [String]
proteins xs = foldr (liftA2 (:))
                    (Just [])
                    (takeWhile (Just "STOP" /=) proteins')
 where
  codons    = chunksOf 3 xs
  proteins' = map translation codons

translation :: String -> Maybe String
translation codon | codon == "AUG"                     = Just "Methionine"
                  | codon == "UGG"                     = Just "Tryptophan"
                  | codon `elem` ["UUU", "UUC"]        = Just "Phenylalanine"
                  | codon `elem` ["UUA", "UUG"]        = Just "Leucine"
                  | codon `elem` ["UCU", "UCC", "UCA", "UCG"] = Just "Serine"
                  | codon `elem` ["UAU", "UAC"]        = Just "Tyrosine"
                  | codon `elem` ["UGU", "UGC"]        = Just "Cysteine"
                  | codon `elem` ["UAA", "UAG", "UGA"] = Just "STOP"
                  | otherwise                          = Nothing
